[1단계 - 로또(자동)] 렉스(오성원) 미션 제출합니다.

Seongwon97 opened the pull request:
안녕하세요 코니 😁

백엔드 4기 교육생 렉스🦖라고 합니다.
벌써 우테코 교육이 시작한지 3주라는 시간이 지났네요.

이번 미션을 진행할때는 TDD, 전략패턴, Enum, 원시값 포장 등 여러 학습을 하며 진행하였습니다.

<hr>

미션을 진행하며 몇가지 궁금한 것이 있어 질문드립니다.
1. **메서드가 public->private으로 변경될 경우 해당 메서드 관련 테스트 메서드는 어떻게 관리를 하는지에 대한 의문**
지난 자동차 경주 미션에 이어 이번 미션도 TDD로 개발을 진행하였습니다. 말단의 기능부터 진행을 하다보니 초기에는 public으로 생성되었던 메서드를 프로그램을 점차 확장하며 private으로 바뀌게 되는 상황들이 종종 발생하고는 하는데 이런 경우 해당 메서드에 대한 테스트는 개발 후반부에는 제거를 해도 될까요??(private로 바뀌는 순간)  <br> 
현재 이번 제출 코드에서 해당하는 메서드는 `PurchasedLotto.java`파일의 `purchase`메서드가 존재합니다. 현재는 해당 메서드를 public으로 남겨두었으나(테스트를 위해 public메서드를 private로 남겨두는 것은 좋지 않은 것은 알고있습니다🥲) 실제 개발을 할 때는 이러한 경우 어떤 식으로 작성한 test코드를 관리하는지 여쭙고 싶습니다🧐

2. **원시값과 문자열 포장의 장점에 대한 의문**
이번 미션을 진행하며 추가된 요구사항에는 `모든 원시 값과 문자열을 포장한다`라는 요구사항이 있었습니다. 해당 요구사항을 지키기 위해 Lotto가 갖는 로또 숫자에 대한 list 원소를 LottoNumber로 변경하였습니다.  <br>
원시값 포장에 대한 이점은 `여러 자료를 찾아봤을 때 해당 값이 무엇을 의미하는지 확실하게 알 수 있다!`, `유효성 검사, 동등성 검사, 불변을 객체 내부에서 할 수 있다!` 라는 장점이 있다고 하더군요. 하지만 이번 미션을 진행하면서는 로또 번호를 int를 사용하지 않고 LottoNumber라는 객체를 만들어 사용하다보니 장점보다는 불편함을 더 크게 느꼈던 것 같습니다. <br>
생각해보면 원시값 포장에 대한 이점인 유효성 검사, 동등성 검사와 같은 것을 제가 작성한 코드에서는 진행하지 않고 유효성 동등성은 모두 해당 객체를 사용하는 Lotto객체에서 진행을 하여 그랬던 것 같습니다. 유효성의 경우 `RandomPurchaseStrategy.java`파일에서 이미 유효한 값들만을 보내주어 필요 없다고 생각하여 진행을 하지 않았습니다..이번 미션에서 원시값 문자열 포장에 대한 이점은 의미있는 객체로 포장한다라는 이점만 느낄 수 있었는데 혹시 추가적인 이점을 느낄 수 있도록 수정할 내용은 따로 없을까요..?

감사합니다.🙇🏻‍♂️

Seongwon97 committed:
docs: 요구사항 정리

Seongwon97 committed:
feat: 로또 넘버 객체 생성

- 같은 번호를 가진 로또 넘버 객체는 같은 객체라고 인지한다.
- readme에 요구사항 추가

Seongwon97 committed:
feat: 로또 넘버 객체 리스트 일급 컬렉션 구현

- 매개변수로 받은 Integer리스트를 옳바르게 저장하는지 테스트 완료

Seongwon97 committed:
feat: 당첨 등수 판별 기능 구현

- 당첨 등수 판별 기능 구현
- 상금 정보를 담은 enum생성

Seongwon97 committed:
feat: 로또 결과 도출 메서드 구현

- 로또 결과를 Map으로 도출하는 메서드 구현
- 전략 패턴을 사용하여 랜덤하게 로또 번호를 생성해주는 메서드 구현

Seongwon97 committed:
feat: 수익률 계산 메서드 구현 완료

- 구매 금액과 상금을 비교하여 수익률을 계산하는 메서드 구현 완료

Seongwon97 committed:
style: formatting 통일

Seongwon97 committed:
refactor: 전략 패키지 파일 분리

- 전략 패키지 관련 파일을 /domain/strategy로 분리

Seongwon97 committed:
refactor: 최종 결과 객체를 일급 컬렉션으로 분리

- PurchadedLotto의 Map객체를 일급 컬렉션으로 분리

Seongwon97 committed:
style: 불필요한 import제거

Seongwon97 committed:
feat: 가격 입력값이 정수가 아닌 경우 예외처리

Seongwon97 committed:
feat: 가격 입력값이 음수인 경우 예외처리

Seongwon97 committed:
refactor: 가격 입력이 1000원 미만일 떄 예외가 발생하도록 수정

- 음의 입력값만 체크하던 메서드를 최소 금액보다 낮을 떄 예외가 발생하도록 수정

Seongwon97 committed:
feat: 6개가 아닌 로또 넘버 입력시 예외처리

Seongwon97 committed:
feat: 로또 번호가 숫자가 아닌 경우 예외처리

Seongwon97 committed:
feat: 로또 번호가 1~45 사이가 아닐 경우 예외처리

Seongwon97 committed:
feat: 중복된 로또 번호 존재시 예외처리

Seongwon97 committed:
feat: 보너스 번호를 이미 갖고 있을 시 예외처리

- 당첨 번호 객체 생성
- 보너스번호와 일반 번호와 중복시 예외처리

Seongwon97 committed:
feat: 로또 구매 기능 구현 완료

- View와 Model의 메서드를 호출하여 연결 완료

Seongwon97 committed:
feat: 당첨번호, 보너스 번호 생성 구현 완료

- 보너스 번호 중복시 반복 입력 구현

Seongwon97 committed:
fix: 3개 미만 일치시 발생하는 오류 수정

Seongwon97 committed:
feat: 최종 상금 통계 출력 기능 구현

Seongwon97 committed:
fix: 최종 상금 통계 출력 순서 수정

- 최종 상금 출력을 5등부터 1등까지 순차적으로 출력되도록 수정

Seongwon97 committed:
feat: 최종 수익률 출력 기능 구현

Seongwon97 committed:
refacter: 매직넘버를 상수로 변경, 불필요한 import제거

- 파일들에 있는 매직넘버들을 모두 상수로 변경
- 불필요한 import들을 제거
- 파일들의 Indent를 통일

Seongwon97 committed:
refactor: 우승 상금 Enum클래스 내부 로직 변경

- WinnerPrice.java -> Prize.java로 이름 변경
- Enum 필드에 bonus정보 추가
- 등수 찾는 로직 개선

Seongwon97 committed:
refactor: 변수, 메서드 명을 성격에 맞게 상금을 prize로 변경

Seongwon97 committed:
refactor: 테스트의 displayName설정

- InputValidationTest.java의 테스트들을 nested 테스트로 묶음
- 테스트 클래스들에 DisplayName추가

Seongwon97 committed:
style: 컨벤션 통일

choihz commented on README.md:
```diff
@@ -2,6 +2,36 @@
+- [ ] 당첨 통계 부분 구현
+  - [ ] 당첨 상금을 세팅한다.
+  - [ ] 로또 번호별 등수를 계산한다.
+  - [ ] 수익률을 계산한다.
```
❓❓

Seongwon97 replied:
조금이나마 더 살아있는 문서를 만들고 처음 생각한 요구사항들을 놓치지 않기 위해 요구사항 정리를 체크리스트로 구성하였는데..구현을 하다보니 체크리스트 관리를 아예 안하게 된거 같네요 😅

현재는 모든 내용을 구현한 상태이기에 체크 리스트 자체를 제거했습니다. 😅

choihz commented on src/main/java/utils/InputValidation.java:
```diff
@@ -0,0 +1,93 @@
+public class InputValidation {
+
+    private static final int LOTTO_MINIMUN_NUMBER = 1;
+    private static final int LOTTO_MAXIMUN_NUMBER = 45;
```
오타가 있네요~

Seongwon97 replied:
앗...오타가 있었네요ㅠ바로 수정했습니다.
다른 코멘트에서 언급해주신 오타, 잘못된 문법/맞춤법들도 모두 수정했습니다.

IntelliJ에서 잘못된 이름들은 `Typo`메시지를 띄워주는군요 😅
intelliJ에서 밑줄로 알려주는 메시지를 넘기지 않고 확인하는 습관을 길러야할거 같네요!!
감사합니다 😁

choihz commented on src/main/java/utils/InputValidation.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    public static List<Integer> validateWinningNumber(final String inputNumbers) {
+        final List<String> splitedNumbers = Arrays.asList(inputNumbers.split(LOTTO_NUMBER_DELIMITER));
```
```suggestion
        final List<String> splitNumbers = Arrays.asList(inputNumbers.split(LOTTO_NUMBER_DELIMITER));
```

<img width="1041" alt="스크린샷 2022-02-24 오후 10 40 06" src="https://user-images.githubusercontent.com/18049492/155535049-ebe2870b-f5b0-45e8-9077-a7d1fb07fa49.png">

Seongwon97 replied:
해당 메서드명의 문법적 오류도 올바르게 수정했습니다!!

choihz commented on src/main/java/utils/InputValidation.java:
```diff
@@ -0,0 +1,93 @@
+    private static void checkNumbersRange(final List<Integer> numbers) {
+        numbers.stream()
+                .forEach(number -> checkNumberRange(number));
+    }
```
[for-loop 를 Stream.forEach() 로 바꾸지 말아야 할 3가지 이유](https://homoefficio.github.io/2016/06/26/for-loop-%EB%A5%BC-Stream-forEach-%EB%A1%9C-%EB%B0%94%EA%BE%B8%EC%A7%80-%EB%A7%90%EC%95%84%EC%95%BC-%ED%95%A0-3%EA%B0%80%EC%A7%80-%EC%9D%B4%EC%9C%A0/)를 읽어볼까요?

Seongwon97 replied:
보내주신 게시글을 읽어봤습니다. 

가독성 측면에서는 게시글이 16년도에 작성된 만큼 요즘에는 많은 사람들이 stream을 사용하게 되어 가독성에 문제가 되지 않는다고 생각합니다. 
다만 stream이 나름 최신 기술이라 for-loop와 비교를 하였을 때 성능적으로 더 좋거나 동일한 성능을 보일거라 생각하였는데, 더 떨어지는 성능을 보인다니 놀랐습니다..

보내주신 링크 이외에도 forEach와 for-loop에 대한 자료들을 찾아봤는데 forEach는 for-loop보다 primitive타입에서 사용시 성능이 현저히 낮으며 iteration cost와 functionality cost의 합이 충분히 클 때, 비로소 성능이 비슷해진다는 것을 알 수 있었습니다.

- 최신 기술이라고 모든 면에서 좋은건 아니다.
- stream의 forEach는 대체로 for-loop보다 성능이 떨어지며, iteration cost와 functionality cost의 합이 클 때 for-loop와 비슷한 성능을 가진다.

덕분에 새로운 내용을 배워갑니다!!
감사합니다🙇🏻‍♂️

choihz commented on src/main/java/utils/InputValidation.java:
```diff
@@ -0,0 +1,93 @@
+
+    private static void checkDuplicateNumber(final List<Integer> numbers) {
+        Set<Integer> uniqueNumber = numbers.stream()
+                .collect(Collectors.toSet());
```
꼭 스트림을 사용해야만 하나요?

Seongwon97 replied:
List를 Set으로 변경하려면 아래의 코드와 같이 Set 인스턴스를 하나 생성하여 값을 하나씩 넣어줘야 하는줄 알고 이러한 코드를 줄이고자 stream을 사용하였습니다.
```java
        Set<Integer> uniqueNumber = new HashSet<>();
        for (int number: numbers) {
            uniqueNumber.add(number);
        }
```

하지만 intelliJ의 `warning`메시지를 확인 후 HashMap의 생성자를 확인해보니 아래와 같이 collection 인스턴스를 받아 HashSet인스턴스를 만들어주는 생성자가 있더군요👍🏻
생성자 확인 후 바로 HashSet생성자를 사용하는 코드로 변경하였습니다!!
```java
    public HashSet(Collection<? extends E> c) {
        map = new HashMap<>(Math.max((int) (c.size()/.75f) + 1, 16));
        addAll(c);
    }
```

choihz commented on src/main/java/utils/InputValidation.java:
```diff
@@ -0,0 +1,93 @@
+        if (numbers.size() != LOTTO_SIZE) {
+            throw new IllegalArgumentException(ERROR_NUM_OF_BALL);
+        }
+    }
```
메서드 이름이 혼란을 일으킬 수 있어 보이네요. 더 나은 단어를 메서드 안에서 바로 발견할 수 있네요.

Seongwon97 replied:
메서드 네임이 `NumOfNumbers`라니 다시봐도 헷갈리는 네이밍인 것 같네요..! 

이번 미션을 진행하면서 가장 어려웠던 점 중에 하나가 클래스와 변수 네이밍이었던 것 같습니다.
로또번호, 로또, 우승번호, 구매한 로또.. 등등 비슷한 이름의 클래스와 변수를 생성하다보니 개발 중반부에는 개발을 하는 저 자신도 헷갈리더군요🤯

비슷한 클래스, 변수들의 네이밍을 하는 코니만의 팁이 있을까요..??

choihz commented on src/main/java/utils/InputValidation.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    private static void checkNumberRange(final int number) {
+        if (!(number >= LOTTO_MINIMUN_NUMBER && number <= LOTTO_MAXIMUN_NUMBER)) {
```
조건문을 더 이해하기 쉽게 수정할 수 있을 것 같아요.

Seongwon97 replied:
헉..조건을 반대로 하면 되는 것을 어째서 `!`를 붙여가면서 저렇게 했을까요ㅠ

조건을 `number < LOTTO_MINIMUM_NUMBER || LOTTO_MAXIMUM_NUMBER < number` 으로 수정했습니다😅

choihz commented on src/main/java/utils/InputValidation.java:
```diff
@@ -0,0 +1,93 @@
+    private static final String ERROR_NUM_OF_BALL = "로또 번호는 6개의 번호를 입력해줘야 합니다.";
+    private static final String ERROR_BALL_NON_INTEGER = "로또 번호는 숫자만 입력해줘야 합니다.";
+    private static final String ERROR_NUMBER_OUT_RANGE = "로또 번호는 1에서 45 사이의 값을 입력해줘야 합니다.";
+    private static final String ERROR_DUPLICATE_NUMBER = "로또 번호는 중복되면 안 됩니다.";
```
너무 많은 검증이 여기에 몰려 있는 것 같아요. `LottoNumber` 등이 자체적으로 검증을 할 수 있게 된다면 어떤 장점이 있을까요?

Seongwon97 replied:
바로 아래의 [코멘트](https://github.com/woowacourse/java-lotto/pull/390#discussion_r813903720) 내용 반영과 함께 로또 번호의 범위 체크, 번호의 중복 체크와 같은 검증 메서드를 진짜 검증이 필요한 위치에 옮겼습니다.

이제 InputValidation.java파일에는 로또 번호 객체와는 크게 연관없는 `값이 정수인지`, `투입금액이 1000원 미만인지`, `6개의 정수가 입력되었는지` 검증하는 메서드 만을 남겨두었습니다.😆

choihz commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -0,0 +1,37 @@
+
+    public LottoNumber(int number) {
+        this.number = number;
+    }
```
바로 위 코멘트와 이어지는 내용입니다. `InputValidation`을 거치지 않으면, 100이든 1000이든 어떤 숫자를 넣어도 `LottoNumber` 객체를 만들 수 있겠네요.

> 이번 미션을 진행하며 추가된 요구사항에는 `모든 원시 값과 문자열을 포장한다`라는 요구사항이 있었습니다. 해당 요구사항을 지키기 위해 Lotto가 갖는 로또 숫자에 대한 list 원소를 LottoNumber로 변경하였습니다.
>
> 원시값 포장에 대한 이점은 `여러 자료를 찾아봤을 때 해당 값이 무엇을 의미하는지 확실하게 알 수 있다!`, `유효성 검사, 동등성 검사, 불변을 객체 내부에서 할 수 있다!` 라는 장점이 있다고 하더군요. 하지만 이번 미션을 진행하면서는 로또 번호를 int를 사용하지 않고 LottoNumber라는 객체를 만들어 사용하다보니 장점보다는 불편함을 더 크게 느꼈던 것 같습니다.
>
> 생각해보면 원시값 포장에 대한 이점인 유효성 검사, 동등성 검사와 같은 것을 제가 작성한 코드에서는 진행하지 않고 유효성 동등성은 모두 해당 객체를 사용하는 Lotto객체에서 진행을 하여 그랬던 것 같습니다. 유효성의 경우 `RandomPurchaseStrategy.java`파일에서 이미 유효한 값들만을 보내주어 필요 없다고 생각하여 진행을 하지 않았습니다..이번 미션에서 원시값 문자열 포장에 대한 이점은 의미있는 객체로 포장한다라는 이점만 느낄 수 있었는데 혹시 추가적인 이점을 느낄 수 있도록 수정할 내용은 따로 없을까요..?

이 질문에 대해서도 약간의(?) 답이 될 수 있겠네요.

Seongwon97 replied:
그렇군요..💡
질문을 할 때는 우승번호를 유저가 직접 입력하는 것은 생각하지 않고, 프로그램이 만들어준 1~45사이의 수로 LottoNumber를 생성하는 것만 생각하여 `굳이 1~45사이의 검증된 수를 받는데 LottoNumber 내부에서 체크를 할 필요가 있을까!?` 라는 생각을 하였습니다. 

다시 생각해보니 프로그램 내에서 유저가 입력해주는 우승번호를 검증해야하는 일이 있었네요. 그리고 무엇보다 유저의 입력이 아니어도 LottoNumber는 1~45사이의 숫자만 가능하다는 것을 LottoNumber내에서 유효성 체크를 해줘야지 발생할 수 있는 오류를 예방할 수 있을 것 같습니다.

객체 포장의 이점!! 이제야 이해했습니다😁

choihz commented on src/main/java/controller/LottoGameController.java:
```diff
@@ -0,0 +1,36 @@
+public class LottoGameController {
+
+    private PurchasedLotto purchasedLotto;
+    private WinningNumber winningNumber;
```
컨트롤러가 구입한 로또와 당첨 번호를 인스턴스 변수로 가지고 있군요.

1. 컨트롤러의 역할이 무엇인지
2. 클래스, 객체, 인스턴스 변수의 관계란 어떤 것인지

에 대해 고민해 보면 좋을 것 같아요.

Seongwon97 replied:
1. 컨트롤러의 역할은 Model과 View사이를 연결해주는 인터페이스 역할이라고 학습을 하였습니다. 각각의 메서드를 호출하여 연결하여 유저가 사용할 수 있는 최종적인 메서드를 생성하는 것이 컨트롤러의 위치라고 생각합니다.
2. 컨트롤러가 인스턴스 변수를 갖고있다보니 컨트롤러와 Model의 관계가 연결된 것이 아니라, 컨트롤러 안에 Model이 들어가있는 형태가 만들어졌던것 같습니다. 인스턴스 변수를 없애며 둘 사이의 관계를 조정해봤습니다!!

choihz commented on src/main/java/domain/PurchasedLotto.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class PurchasedLotto {
```
실제로 로또를 발행하고, 그것을 저장하고, 결과와 수익률까지 계산하는 다재다능한 친구로군요. 하지만, 너무 많은 역할을 몰아준 건 아닐까요?

Seongwon97 replied:
한 객체에서 너무 많은 일을 하는 것 같아 PurchasedLotto에서 하던 일들을 Store에서 로또의 구매, PrizeResult에서는 결과와 수익률 계산, Lottos에서 발행된 로또의 저장을 하도록 작업을 나눠봤습니다!!😁

choihz commented on src/main/java/utils/InputValidation.java:
```diff
@@ -0,0 +1,93 @@
+    private static final String ERROR_NUMBER_OUT_RANGE = "로또 번호는 1에서 45 사이의 값을 입력해줘야 합니다.";
+    private static final String ERROR_DUPLICATE_NUMBER = "로또 번호는 중복되면 안 됩니다.";
+
+    public static int validatePrice(final String inputPrice) {
+        final int price = checkNonInteger(inputPrice, ERROR_PRICE_NON_INTEGER);
+        checkUnderMinimumPrice(price);
+
+        return price;
+    }
```
`final`을 사용하는 본인만의 기준이 있으신가요? 전체적으로 일관성이 있으면 더 좋을 것 같아요.

choihz commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -0,0 +1,37 @@
+
+public class LottoNumber {
+
+    private int number;
```
`final` 키워드의 유무가 어떤 차이를 만들어낼까요?

Seongwon97 replied:
`final`키워드는 정확히 불변! 이라기보다 재할당을 금지한다는 표현이 맞다고 생각합니다.

final을 통해 재할당을 금지함으로써 프로그램 내에서 개발자로 실수로 객체를 새로 재할당 하는 것을 컴파일 시점, 또는 그 이전에 막을 수 있다고 생각합니다.

## 이번 미션을 진행하며 저만의 final 키워드 기준을 세워봤습니다.
### 클래스 변수
가능하면 final을 붙이자.
### 인스턴스 변수
생성자 내부에서 초기화 할 수 있으면 final을 붙이자.
### 메서드의 매개변수
매개변수에 붙는 final의 경우 원시타입은 재할당을 막을 수 있지만, 그 외의 타입의 경우 내부 필드값의 변경을 막지 못한다. 그리하여 원시타입에는 final을 붙이고 그 외에는 붙이지 않는다.
### 메서드 내부의 변수
재할당을 금지할 변수의 경우 final을 붙이자.

choihz commented on src/main/java/domain/Prize.java:
```diff
@@ -0,0 +1,48 @@
+    THIRD(1500000, 5, false),
+    FOURTH(50000, 4, false),
+    FIFTH(5000, 3, false),
+    FAIL(0, 0, false);
```
아무래도 0이 많아지면 읽기 어렵고, 실수할 가능성도 높아지겠죠? [Underscores in Numeric Literals](https://docs.oracle.com/javase/7/docs/technotes/guides/language/underscores-literals.html) 문서를 읽고 적용해 보아요!

Seongwon97 replied:
덕분에 새로운 정보를 알아갑니다. 숫자 단위가 너무 높아 미션을 진행하면서도 0의 개수를 올바르게 입력하였는지 여러번 확인하곤 하였는데 앞으로 높은 단위의 숫자들은 underscore(_)를 잘 활용해야겠네요😁

choihz commented on src/main/java/domain/PrizeResult.java:
```diff
@@ -0,0 +1,43 @@
+                .filter(winnerPrice -> winnerPrice != Prize.FAIL)
+                .sorted(Comparator.comparing(Prize::getPrize))
+                .collect(Collectors.toList());
+    }
```
메서드 이름만 보고 메서드가 하는 일을 잘 파악할 수 있을까요?

Seongwon97 replied:
코니의 리뷰를 보고 코드를 확인해보니, 코드를 작성한 저 마저도 호출하는 부분의 코드를 확인하기 전에는 해당 메서드의 이름만 보고 어디에 쓰이는 메서드인지 알아차리기 힘들었던 것 같네요. 

키 값을 우승 상금을 기준으로 정렬한 키들의 리스트를 반환하여 `sortedPriceKeySet`으로 수정하였습니👍🏻

choihz commented on src/main/java/domain/PurchasedLotto.java:
```diff
@@ -0,0 +1,57 @@
+    public void purchase(PurchaseStrategy purchaseStrategy) {
+        Lotto lotto = new Lotto(purchaseStrategy.getNumbers());
+        lottos.add(lotto);
+    }
```
> 지난 자동차 경주 미션에 이어 이번 미션도 TDD로 개발을 진행하였습니다. 말단의 기능부터 진행을 하다보니 초기에는 public으로 생성되었던 메서드를 프로그램을 점차 확장하며 private으로 바뀌게 되는 상황들이 종종 발생하고는 하는데 이런 경우 해당 메서드에 대한 테스트는 개발 후반부에는 제거를 해도 될까요??(private로 바뀌는 순간)
>
> 현재 이번 제출 코드에서 해당하는 메서드는 `PurchasedLotto.java`파일의 `purchase`메서드가 존재합니다. 현재는 해당 메서드를 public으로 남겨두었으나(테스트를 위해 public메서드를 private로 남겨두는 것은 좋지 않은 것은 알고있습니다🥲) 실제 개발을 할 때는 이러한 경우 어떤 식으로 작성한 test코드를 관리하는지 여쭙고 싶습니다🧐

`private` 메서드라면 결국 이걸 직/간접적으로 사용하는 `public` 메서드가 있을테니 그것으로 테스트할 수 있겠죠. 만약 테스트하기 어렵다면 어떤 부분에서 어려운 것인지, 해결할 수 있는 방법이 없을지 고민해 봐야 하겠죠?

또 생각해 볼 지점이 있다면, 테스트 코드를 작성하는 목적에 관한 것입니다. 테스트 코드를 왜 작성할까요? 메서드 하나하나가 제대로 동작하는지 검증하기 위해서? 그렇다면 모든 메서드를 `public`으로 열어 자유롭게 테스트하는 것이 역시 가장 좋은 방법 아닐까요? 왜 이런 방식이 좋지 않다고 다들 말하는 걸까요?

Seongwon97 replied:
접근 제어자의 장점으로 객체의 내부 로직을 보호하기 위해 외부의 접근을 허용/차단하여 객체의 오용을 막고 사용자들이 객체에 집중할 수 있도록 돕는다는 장점이 있습니다. 모든 메서드를 테스트하기 위해 접근 제어자를 private에서 public으로 변경할 경우 위의 접근 제어자의 장점을 모두 잃게 되어 테스트를 통해 얻는 이점보다 손해가 더 클 것으로 생각됩니다. 그렇기에 코니가 언급해주신대로 private메서드를 테스트 하기 위해서는 직/간접적으로 해당 메서드를 사용하는 public메서드를 테스트해야하는 것 같습니다. 

저도 코니가 언급해주신대로 purchase 메서드를 테스트하기 위해 purchase를 호출하여 사용하는 public메서드를 통해 올바른 개수의 로또를 구매하였는지 테스트하도록 변경해봤습니다!!

choihz commented on src/main/java/domain/WinningNumber.java:
```diff
@@ -0,0 +1,34 @@
+    private static final String ERROR_DUPLICATE_BONUS = "보너스 번호가 당첨 번호와 중복됩니다.";
+
+    private List<LottoNumber> winningNumbers;
+    private LottoNumber bonus;
```
`LottoNumber`가 나타내는 것은 로또 숫자 한 개지만, `WinningNumber`는 완전히 다르군요. 이름이 비슷하기 때문에 헷갈릴 수도 있을 것 같아요.

Seongwon97 replied:
개발을 하는 저의 입장에서는 `LottoNumber`는 로또번호가 될 수 있는 수!, `WinningNumber`는 당첨번호! 를 생각하여 작성하였는데 코드 작성자가 아닌 제 3자의 입장에서는 헷갈리게 보일 수 있었군요😅 

헷갈림을 방지하고자 `LottoNumber`는 `Lotto`로 `WinningNumber`는 `WinningNumbers`로 변경을 해보았습니다.

choihz commented on src/test/java/domain/PurchasedLottosTest.java:
```diff
@@ -0,0 +1,55 @@
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+@DisplayName("상금과 수익률을 옳바르게 계산하는지 테스트한다.")
```
```suggestion
@DisplayName("상금과 수익률을 올바르게 계산하는지 테스트한다.")
```

다만, 여기에까지 이름이 있을 필요가 있는지 의문이긴 하네요. 각 테스트 메서드에 이름을 잘 붙여주는 것이 더 좋지 않을까요?

Seongwon97 replied:
일단 오타는 수정했습니다!!

테스트 클래스 별로 붙이는 `@DisplayName`의 경우 초기에는 붙이지 않았으나 클래스별로 각각 어떤 테스트 내용들이 있는지 알리고자 붙여봤습니다.

각각의 파일명에서 `@DisplayName`을 붙여두니 어떤 테스트들이 존재하는지 확실히 알 수 있어서 더 좋다고 생각하였습니다. 하지만 더 깊게 생각해보니 현재는 미션의 규모가 작아 테스트 오류가 발생하여도 어떤 파일에서 오류가 발생하였는지 쉽게 찾을 수 있는데 프로젝트 규모가 커져서 테스트 파일만 100개이상 존재하게 된다면 재정의한 `@DisplayName`으로 어떤 파일에서 테스트 실패가 나타나는지 확인하기 힘들거같다는 생각이 드네요😅

실무에서는 보통 테스트 클래스 위에는 DisplayName을 붙이지 않고 테스트 메서드 위에만 붙이는 편인가요?? 
InputValidationTest.java파일에서 사용한 Nested Test의 경우도 붙이지 않는지 궁금합니다🧐

choihz commented on src/test/java/domain/PurchasedLottosTest.java:
```diff
@@ -0,0 +1,55 @@
+    void init() {
+        inputMoney = 1400;
+        lottos = new PurchasedLotto(inputMoney);
+        lottos.purchase(new CustomPurchaseStrategy(Arrays.asList(new Integer[]{1, 2, 3, 43, 44, 45})));
```
```suggestion
        lottos.purchase(new CustomPurchaseStrategy(Arrays.asList(1, 2, 3, 43, 44, 45)));
```

IntelliJ를 사용하고 있다면 조금 더 주의를 기울여 볼까요?

Seongwon97 replied:
해당 코드도 `Redundant array creation for calling varargs method` warning메시지를 띄워줬네요 😅

앞으로 intelliJ가 주는 메시지들에 더욱 신경을 써보겠습니다🔥

choihz commented on src/test/java/domain/PurchasedLottosTest.java:
```diff
@@ -0,0 +1,55 @@
+
+    PurchasedLotto lottos;
+    WinningNumber winningLotto;
+    int inputMoney;
```
테스트 코드에서도 접근 제어자를 확인해 주세요.

Seongwon97 replied:
테스트 코드에 접근 제어자 추가했습니다!!

choihz commented on src/test/java/domain/PurchasedLottosTest.java:
```diff
@@ -0,0 +1,55 @@
+
+    @Test
+    @DisplayName("옳바른 수익률이 계산된다.")
+    void lottos_calulateEarningRate() {
```
(중요하진 않지만) 오타가 있네요~ 역시 IntelliJ가 잡아주고 있기도 하고요.

Seongwon97 replied:
해당 오타도 수정했습니다!!
저 자신보다는 intelliJ를 더 믿어야 할거 같네요..🤯

choihz commented on src/main/java/utils/InputValidation.java:
```diff
@@ -0,0 +1,93 @@
+        return price;
+    }
+
+    public static List<Integer> validateWinningNumber(final String inputNumbers) {
```
<img width="342" alt="스크린샷 2022-02-24 오후 11 43 11" src="https://user-images.githubusercontent.com/18049492/155546181-7ee40180-7a66-4051-ba31-1b219921243e.png">

🤔

Seongwon97 replied:
미션 예시에 `1, 2, 3, 4, 5, 6`과 같이 나와있어 구분자를 `, `로 정의해뒀는데 공백이 없는 상황에서도 분리가 되도록 설정하면 더 좋겠군요.

공백을 포함한 ", " 와 ","으로 모두 구분되도록 수정했습니다😁

choihz reviewed (changes requested):
안녕하세요, 렉스! 로또 미션을 함께할 리뷰어 코니입니다 👋

로또 미션 1단계를 잘 구현해 주셨네요 👍 몇 가지 코멘트 남겼는데 확인 부탁드리고, 이외에도 더 궁금한 점이나 의견이 있으시다면 언제든 코멘트 또는 DM 주세요~

Seongwon97 committed:
docs: 요구사항 정리 체크리스트 제거

Seongwon97 committed:
refactor: 잘못된 문법, 오타로 작성된 이름들 수정

- InputValidation.java
  - MAXIMUN -> MANIMUM
  - MINIMUN -> MINIMUM
  - splitedNumbers -> splitNumbers
- InputValidation.java, LottoTest.java
  - 옳바른 -> 올바른
- PurchasedLottosTest.java
  - 옳바른 -> 올바른
  - calulateEarningRate -> calculateEarningRate

Seongwon97 committed:
refactor: 테스트 클래스의 DisplayName제거

- 테스트 파일의 수가 많아지게 된다면 DisplayName은
어떤 테스트에서 오류가 발생하였는지 알기 힘들어지게 만들 것으로 판단하여 제거

Seongwon97 committed:
refactor: 오타수정, 리스트 생성 매개변수 수정

- RandomPurchaseStrategy.java
  - MAXIMUN -> MANIMUM
  - MINIMUN -> MINIMUM
- PurchasedLottosTest.java
  - 리스트 생성 매개변수 수정하여 warning메시지 제거

Seongwon97 committed:
refactor: checkDuplicateNumber의 Set인스턴스를 생성자를 활용하도록 수정

- stream을 사용하여 list를 set으로 변경하던 코드를 HashSet생성자를 통해 변경하도록 수정

Seongwon97 committed:
refactor: 로또 번호의 범위 조건과 애매한 매서드 명 수정

- checkNumberRange에 있는 로또 번호의 범위 조건을 가독성 있게 수정
- checkNumOfNumbers메서드 명을 checkNumOfBalls로 더욱 의미있게 수정

Seongwon97 committed:
refactor: 로또 넘버 범위 체크, 중복 체크 메서드의 위치 변경

- InputValidation.java에서 하던 로또 넘버의 범위 체크, 중복 체크 메서드를
LottoNumber.java, WinningNumber.java로 이동
- main의 메서드 위치 이동에 따른 test메서드 이동
- Lotto.java의 로또 번호를 저장하는 List를 Set으로 변경

Seongwon97 committed:
refactor: 테스트 코드의 접근 제어자 수정

Seongwon97 committed:
refactor: 컨트롤러의 인스턴스 변수 제거

- 컨트롤러가 갖는 인스턴스 변수를 제거하고 Model과 View의 메서드를
호출하고 연결해주는 역할을 하도록 변경
- PrizeResult.java의 불필요한 stream제거

Seongwon97 committed:
refactor: 최종 결과 도출 관련 메서드를 PrizeResult로 이동

- 구매한 로또 정보를 가진 PurchadedLotto에 있던 로또 당첨 통계 관련 메서드를
모두 로또 결과 정보를 담는 PrizeResult로 이동

Seongwon97 committed:
refactor: 인스턴스 필드에 final 키워드 추가

- 값의 변경이 없을 인스턴스 필드에 대해 final 키워드를 추가
- Lotto.java의 matchedRegularNumbers메서드 로직을 더욱 간단하게 변경

Seongwon97 committed:
refactor: 로또 번호 입력 구분자 수정

- 로또 번호의 구분자를 ", "와 공백이 없는 "," 모두 가능하도록 수정

Seongwon97 committed:
refactor: 메서드, 변수들 이름 수정

Seongwon97 committed:
refactor: View와 PurchasedLotto객체의 결합성 제거

- PurchasedLotto를 매개변수로 받아 구매 로또 들을 출력하던
printPurchasedLotto객체를 Lotto객체의 리스트를 받도록 수정
- Lotto, PurchasedLotto의 getter를 read-only객체로 반환하도록 수정

Seongwon97 committed:
refactor: 정렬된 상금 키 값을 반환하는 메서드 명 변경

- PrizeResult의 유효하며 정렬된 상금으로 정렬된 key값을 반환하는 메서드 명을
validWinnerPrices()에서 sortedPriceKeySet()으로 변경

Seongwon97 committed:
refactor: 헷갈리는 클래스명 변경

- Prize -> Rank
- LottoNumber -> Number
- PurchasedLotto -> Lottos

Seongwon97 committed:
refactor: Lottos의 작업 분리

- Lottos의 로또 구매 로직을 Store클래스를 생성하여 작업 분리

Seongwon97 committed:
test: main리펙터링에 따른 테스트 위치 조정

Seongwon97 committed:
refactor: 매직넘버를 상수로 변환

- 보너스 번호 확인 여부를 체크하는 기준인 5를 상수로 변경

Seongwon97 committed:
refactor: WinnerNumber클래스를 WinnerNumbers로 변경

- 우승번호들의 의미를 갖는 것을 표현하기 위해 WinnerNumber를 WinnerNumbers로 변경

Seongwon97 committed:
refactor: 나만의 final기준에 맞게 final키워드 추가

-클래스 변수: 가능하면 final을 붙이자.

-인스턴스 변수: 생성자 내부에서 초기화 할 수 있으면 final을 붙이자.

-메서드의 매개변수: 매개변수에 붙는 final의 경우 원시타입은 재할당을 막을 수 있지만,
그 외의 타입의 경우 내부 필드값의 변경을 막지 못한다.
그리하여 원시타입에는 final을 붙이고 그 외에는 붙이지 않는다.

-메서드 내부의 변수: 재할당을 금지할 변수의 경우 final을 붙이자.

Seongwon97 committed:
refactor: 상금 금액에 underscore를 추가해 가독성 향상

- 높은 단위의 상금 금액에 underscore(_)를 추가하여 코드의 가독성 향상

Seongwon97 committed:
refactor: PrizeResult의 수익률 변수 제거

- 수익률 필드인 earningRate를 제거 후 메서드를 통해 바로 반환하도록 변경
- earningRate를 제거하며 PrizeResult를 일급 컬렉션으로 변경

Seongwon97 committed:
style: 컨벤션 통일

Seongwon97 committed:
refactor: InputValidation 메서드 위치 조정

Seongwon97 commented:
안녕하세요 코니 👋🏻

보내주신 코멘트들을 보고 코드를 수정해봤어요!

리펙터링을 하며 주로 신경쓴 부분 다음과 같습니다.
- 너무 많은 일을 하는 클래스들의 업무 분배
- Controller가 Model과 View의 인터페이스 역할만을 하도록 변경
- final을 사용하는 나만의 기준 만들기
- 메서드, 변수, 클래스의 네이밍 수정

특히 클래스들의 업무 분배를 하며 코드 구조를 변경하다보니 많은 시간들이 걸렸던 것 같습니다😭
그래도 코드를 리펙터링하고 보니 확실히 이전 코드보다는 발전된 코드가 된 것 같습니다! 

<hr>

리펙터링을 진행하면서도 몇가지 궁금한 내용들이 생겼습니다🧐
### 1. TDD개발론의 리펙터링
이번 미션도 TDD개발론으로 개발을 하며 Test코드를 먼저 작성하고 Main로직을 작성하는 red-green-refactor의 순서로 진행을 하였습니다. 지난 미션에서는 전체적인 구조를 변경하지 않아 리펙터링에서 큰 문제가 없었는데, 이번 미션에서는 리펙터링으로 클래스들의 업무 분배를 하며 새로운 클래스를 생성하는 등의 코드 로직의 위치 변경을 많이 하다보니 `리펙터링도 TDD로 test코드 위치를 먼저 옮기고 main을 고쳐야하나..?`라는 살짝 바보처럼 보일 수 있는 생각이 들더군요😂 
이번 리펙터링을 진행하며 저같은 경우는 main을 먼저 수정하고 test코드의 로직을 올바른 위치로 옮기는  순서로 진행을 하였으나 TDD에서도 이것이 일반적인 리펙터링 순서가 맞는지 여쭙고 싶습니다.

### 2. 코니의 `final` 사용 기준
이번 리펙터링을 진행하며 `final`을 사용하는 저만의 기준을 세워봤습니다.  
> - 클래스 변수: 가능하면 final을 붙이자.
> - 인스턴스 변수: 생성자 내부에서 초기화 할 수 있으면 final을 붙이자.
> - 메서드의 매개변수: 매개변수에 붙는 final의 경우 원시타입은 재할당을 막을 수 있지만, 그 외의 타입의 경우 내부 필드값의 변경을 막지 못한다. 그리하여 원시타입에는 final을 붙이고 그 외에는 붙이지 않는다.
> - 메서드 내부의 변수: 재할당을 금지할 변수의 경우 final을 붙이자.

기준을 세웠지만 메서드 내에서는 여기도 과연 `final`을 사용하면 좋을까..?하는 생각들이 많이 들곤합니다. 특히 메서드의 길이가 10줄도 안되는 이번 미션 코드의 경우, 오히려 `final`을 메서드의 매개변수와 변수에 붙이면 오히려 코드가 덜 깔끔해보이는??느낌이 들곤 했습니다.

혹시 메서드를 작성할 때, 코니의 `final`사용 기준을 공유해주실 수 있으신가요?

choihz commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,63 @@
+    public static void printPurchasedLotto(List<Lotto> lottos) {
+        System.out.println(lottos.size() + PURCHASE_COUNT_MESSAGE);
+        for (Lotto lotto : lottos) {
+            System.out.print(LEFT_BRACKET);
+            String lottoNumbers = String.join(DELIMITER, lotto.getLottoNumbers().stream()
+                    .map(lottoNumber -> String.valueOf(lottoNumber.getNumber()))
+                    .collect(Collectors.toList()));
+            System.out.print(lottoNumbers);
+            System.out.println(RIGHT_BRACKET);
```
```suggestion
            String lottoNumbers = lotto.getLottoNumbers().stream()
                    .map(lottoNumber -> String.valueOf(lottoNumber.getNumber()))
                    .collect(Collectors.joining(DELIMITER, LEFT_BRACKET, RIGHT_BRACKET));
            System.out.print(lottoNumbers);
```

이렇게도 가능할 것 같아요~

Seongwon97 replied:
Collectors의 joining은 prefix와 suffix의 설정도 가능했군요!
2단계 코드에 반영했습니다!

choihz commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,43 @@
+
+    private final Set<Number> lottoNumbers;
+
+    public Lotto(List<Integer> numbers) {
+        lottoNumbers = new HashSet<>();
+        for (Integer number : numbers) {
+            lottoNumbers.add(new Number(number));
+        }
+    }
```
생성자로 들어오는 리스트에 숫자가 한 개 있다면, 숫자 하나로 로또를 생성하겠네요. 또는 리스트에 숫자가 여섯 개 들어오지만 같은 게 두 개 있다면, 숫자 다섯 개로 로또를 생성하겠네요. 그렇다면 로또 객체를 사용하는 입장에서, 이 객체가 항상 완전하다고 보장할 수 있을까요?

Seongwon97 replied:
작성한 프로그램 내에서는 중복된 값이 없는 6개의 숫자가 들어가도록 하였으나, 로또 객체의 입장에서는 해당 객체가 6개의 중복되지 않는 숫자가 올바르게 들어와서 생성된다는 것을 보장할 수가 없었네요..🥲 

해당 내용을 보장하기 위해 Number를 저장하는 변수 타입을 List에서 Set으로 변경 후 생성자에서 6개의 숫자가 올바르게 들어왔는지 체크하는 식으로 변경해봤습니다

choihz commented on src/main/java/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,46 @@
+        checkDuplicateNumber(winningNumbers);
+    }
+
+    private void checkDuplicateNumber(final Set<Number> numbers) {
+        if (numbers.size() != LOTTO_SIZE) {
+            throw new IllegalArgumentException(ERROR_DUPLICATE_NUMBER);
+        }
+    }
```
메서드 이름은 `checkDuplicateNumber()`지만, 과연 중복만 체크하게 될까요? 만약 위에 있는 생성자에 엘리먼트가 하나인 리스트가 들어온다면?

Seongwon97 replied:
Lotto와 WinningNumbers에서 Number를 저장하는 변수의 타입을 List에서 Set으로 변경하며 해당 메서드명을 `checkRightSize`으로 변경하여 Set의 size를 체크하는 형식으로 변경했습니다!!

choihz commented on src/main/java/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,46 @@
+        }
+    }
+
+    public void addBonusNumber(final int number) {
```
컨트롤러의 `initWinningNumber()`에서 사용하는 것처럼, 이 객체는 이 메서드가 호출되어야 비로소 완전한 우승 로또가 되네요. 따라서 이 객체를 사용하는 곳에서 실수로 이 메서드 호출을 빼먹으면, 보너스 볼이 없는 우승 로또가 되죠.

보너스 볼이 없는 우승 로또가 생성될 가능성이 있다면, 생성된 `WinningNumbers` 객체를 믿고 사용할 수 있을까요? 객체를 사용할 때마다 매번 `bonus`가 `null`이 아닌지 체크해야 하지 않을까요?

Seongwon97 replied:
보너스 넘버를 입력하였을 때 중복되는 수인가 체크를 하는 검증을 넣기 위해 따로 분리를 하였는데.. 부작용으로 불안정한 `WinnerNumbers`객체가 생성될 수 있다는 문제가 생겼네요..

일반 당첨번호와 보너스 번호를 생성자에서 한번에 넣어주는 방식으로 변경하였습니다.
그대신 당첨번호와 보너스 번호의 예외처리가 당첨번호와 보너스 번호가 모두 입력된 이후에 된다는 손실?을 보게 된거 같아요🥲

choihz commented on src/main/java/utils/InputValidation.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class InputValidation {
```
이 클래스는 `utils` 패키지에 위치해 있네요.
보통 유틸이라는 단어는, 비즈니스 로직과 직접적인 관계가 없고 범용적으로 사용될 수 있는 도구들을 가리켜요. 예를 들어 Java의 `java.util` 패키지를 살펴보면 좋을 것 같아요.

그런데 `InputValidation`은 어떤가요? 무슨 역할을 담당하고 있나요?

Seongwon97 replied:
지금까지 검증 로직이 util에 위치하는 것에 대해 의문을 가진 적이 한번도 없었는데, 코니의 피드백을 들어보니 그동안 잘못된 프로그래밍을 했던 것 같네요..😭

저의 코드를 보니 `입력 금액이 1000원 이하일 경우 예외상황 throw`, `당첨번호의 입력값이 6개 이하일 경우 예외상황 throw`의 경우는 확실히 비즈니스 로직과 직접적인 관계를 갖고 있었고 `각각의 입력값을 integer 또는 Set<Integer>으로 변경하여 값을 반환`도 어떻게 보면 view에서 입력을 받는 작업 뿐만 아니라 비즈니스 로직에서 실제로 사용할 값들을 가공하는 일을 하고있다고 볼 수 있다고 느꼈어요.

해당 내용들은 범용적으로 사용되는 것이 아닌 특정 상황에 사용되는 비즈니스 로직이라서 각각의 검증이 필요한 domain으로 검증 메서드들을 이동시켰습니다!!😎

choihz commented on src/main/java/domain/Lottos.java:
```diff
@@ -0,0 +1,25 @@
+
+    private final List<Lotto> lottos = new ArrayList<>();
+
+    public Lottos(List<Lotto> lottos) {
+        for (Lotto lotto : lottos) {
+            this.lottos.add(lotto);
+        }
+    }
```
여기에도 IntelliJ의 suggestion이 뜨네요.
그런데 `Lotto`의 리스트를 받아 새로운 리스트에 다시 저장하고 있군요. 의도하신 바가 있나요?

Seongwon97 replied:
의도한 바는 없었습니다🥲
아직 api들에 대한 학습이 부족한 것 같습니다.

더 노력해서 알아가야겠어요😭

choihz commented on src/main/java/domain/PrizeResult.java:
```diff
@@ -0,0 +1,56 @@
+        return totalPrice;
+    }
+
+    public List<Rank> sortedPriceKeySet() {
+        return prizeResult.keySet().stream()
+                .filter(winnerPrice -> winnerPrice != Rank.FAIL)
+                .sorted(Comparator.comparing(Rank::getPrize))
+                .collect(Collectors.toList());
+    }
```
1. 메서드 이름만 보고 역할을 쉽게 알 수 있나요? 또한, 구현을 지나치게 드러내는 이름은 아닐까요?
2. 여기서 해야 하는 일이 맞을까요?

Seongwon97 replied:
1. 메서드의 이름을 통해 정렬된 key들을 얻는 것은 알 수 있을 것이라 생각하나, 구현에 대한 정보를 지나치게 드러냈던 것 같습니다..
2. 처음에는 key값을 반환하는 것이라 해당 클래스에 존재하는 것이 당연하다고 생각하였습니다. 하지만 코니의 피드백을 받고 생각을 해보니 해당 정보들은 Rank클래스에도 존재하는 정보라 Rank에서 반환을 하는 것이 더 좋을 것 같아 위치 수정을 하였습니다.

choihz commented on src/test/java/domain/NumberTest.java:
```diff
@@ -0,0 +1,31 @@
+public class NumberTest {
+
+    @Test
+    @DisplayName("같은 숫자를 가진 로또 넘버 객체는 동일하다.")
```
**동일**하다는 것과 **동등**하다는 것의 차이를 학습해 볼까요?

Seongwon97 replied:
동일성은 `==`을 통한 주소값 체크이며, 동등성은 `equals()`를 통해 오브젝트의 내부 값을 비교하는 것이었군요! 

학습을 하며 블로그로 정리도 해봤습니다😎
- [동일성과 동등성](https://seongwon97.github.io/posts/%EB%8F%99%EB%93%B1%EC%84%B1%EA%B3%BC-%EB%8F%99%EC%9D%BC%EC%84%B1/)

choihz commented on src/test/java/domain/NumberTest.java:
```diff
@@ -0,0 +1,31 @@
+        assertThat(lottoNumber1).isEqualTo(lottoNumber2);
+    }
+
+    @DisplayName("보너스 번호가 범위를 벗어난 경우 예외가 발생한다.")
+    @ParameterizedTest(name = "{index} {displayName} bonus={0}")
+    @ValueSource(ints = {0, -1, 46})
+    void checkBonusOutRange_throwIllegalException(final int bonus) {
+        assertThatThrownBy(() -> new Number(bonus))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining("로또 번호는 1에서 45 사이의 값을 입력해줘야 합니다.");
+    }
```
이 테스트에 **보너스 번호**라는 맥락이 담겨 있나요?

Seongwon97 replied:
초기에 보너스 번호를 체크하는 용도로 만들어진 테스트 메서드를 리펙터링을 하며 위치만 옮기다보니 메서드의 이름을 확인하는 것을 놓친 것 같습니다..

테스트를 하는 작업에 맞게 "로또 번호가 범위를 벗어난 경우 예외가 발생한다."로 변경해봤습니다!

choihz commented on src/test/java/domain/PrizeResultTest.java:
```diff
@@ -0,0 +1,54 @@
+class PrizeResultTest {
+
+    private PrizeResult prizeResult;
+    private final int inputMoney = 3000;
```
`inputMoney`는 굳이 여기에 뺄 필요가 없어 보이네요.

Seongwon97 replied:
확인해보니 해당 인스턴스 변수는 수익률을 올바르게 계산을 하는지 테스트하는 메서드에서만 사용하고 있었네요.

해당 변수를 메서드 내부로 이동시켰습니다!

choihz commented on src/test/java/domain/LottoTest.java:
```diff
@@ -0,0 +1,71 @@
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+public class LottoTest {
```
> 이번 미션도 TDD개발론으로 개발을 하며 Test코드를 먼저 작성하고 Main로직을 작성하는 red-green-refactor의 순서로 진행을 하였습니다. 지난 미션에서는 전체적인 구조를 변경하지 않아 리펙터링에서 큰 문제가 없었는데, 이번 미션에서는 리펙터링으로 클래스들의 업무 분배를 하며 새로운 클래스를 생성하는 등의 코드 로직의 위치 변경을 많이 하다보니 리펙터링도 TDD로 test코드 위치를 먼저 옮기고 main을 고쳐야하나..?라는 살짝 바보처럼 보일 수 있는 생각이 들더군요😂
> 이번 리펙터링을 진행하며 저같은 경우는 main을 먼저 수정하고 test코드의 로직을 올바른 위치로 옮기는 순서로 진행을 하였으나 TDD에서도 이것이 일반적인 리펙터링 순서가 맞는지 여쭙고 싶습니다.

세상이 그렇게 교과서적으로 돌아가지는 않죠 ㅎㅎ TDD에 너무 집착할 필요는 없다는 말씀을 드리고 싶어요. 프로덕션 코드를 먼저 수정했다고 하더라도, 이를 검증할 수 있는 테스트 코드를 작성하셨다면 충분히 훌륭한 겁니다 👍

리팩터링 과정에서 새로운 클래스를 만들고 책임을 옮기는 등 구조의 변경이 컸다면 테스트 코드를 따라서 수정해야 하는 것은 어쩔 수 없지요. 그렇지만, 프로덕션 코드가 변경될 때마다 테스트 코드가 항상 같이 변경된다면 테스트 코드를 잘 작성한 것이 맞는지 고민할 필요도 있습니다. 테스트 코드가 프로덕션 코드의 구현에 지나치게 의존하고 있다는 신호일 수도 있어요.

**테스트 코드가 프로덕션 코드의 구현에 지나치게 의존하고 있다**는 건 어떤 의미일까요? 왜 이게 문제가 될까요?

choihz commented on src/main/java/controller/LottoGameController.java:
```diff
@@ -0,0 +1,44 @@
+public class LottoGameController {
+
+    public void run() {
+        final int inputMoney = InputView.inputPrice();
+        final Lottos purchasedLotto = Store.purchaseLottos(inputMoney);
```
> 기준을 세웠지만 메서드 내에서는 여기도 과연 `final`을 사용하면 좋을까..?하는 생각들이 많이 들곤합니다. 특히 메서드의 길이가 10줄도 안되는 이번 미션 코드의 경우, 오히려 `final`을 메서드의 매개변수와 변수에 붙이면 오히려 코드가 덜 깔끔해보이는??느낌이 들곤 했습니다.

사실 이런 부분은 프로젝트 전체에 일관성이 있는 것이 가장 좋기 때문에, 기존 프로젝트의 방식이 있다면 그것을 따르는 것이 좋습니다. 하지만 일반적으로 클래스 변수와 인스턴스 변수에 붙이는 `final`은 중요한 의도를 품고 있기 때문에 때에 맞춰 사용하죠. 그 외에 파라미터나 로컬 변수의 경우엔 붙였을 때의 실익이 크지 않은 경우가 많아서 저는 사용하지 않는 편이에요. 우리가 우테코 내에서 지켜야 하는 프로그래밍 요구 사항을 잘 지키고 있다면, 더욱이 그 필요성이 크지 않죠.

choihz commented on src/test/java/domain/PrizeResultTest.java:
```diff
@@ -0,0 +1,54 @@
+        assertThat(earningRate).isEqualTo(10516.66f);
+    }
+
+}
```
[이 스레드](https://stackoverflow.com/questions/5813311/whats-the-significance-of-the-no-newline-at-end-of-file-log)를 읽어보세요 😉

choihz reviewed (approved):
안녕하세요, 렉스!

지난번 리뷰 이후 많은 고민을 하신 것이 느껴지네요 👍 질문에 대한 답과 몇 가지 코멘트를 더 남겨 두었는데, 다음 단계를 진행하면서 함께 보셔도 될 것 같아서 이번엔 마무리할게요.

그럼, 2단계에서 다시 만나요~
